### PR TITLE
Ctest

### PR DESCRIPTION
**Pull request by stephenworsley**
This should highlight everything added so far.

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
+  </component>
+</project>
```
Comment (lbdreyer) [inline .idea/vcs.xml:6]: I suspect this file was created by your IDE.
It is usually best to keep these sort of files out of your git repo either by not ever adding it, i.e. don't do:
```$ git add .idea/vcs.xml```
Or by including the '.idea' directory in a `.gitignore` file.

File: my-WOPI.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
+import argparse
+import json
+import matplotlib.pyplot as plt
+import matplotlib.animation as animation
+import matplotlib
+from math import pi
+import numpy as np
+import datetime
```
Comment (lbdreyer) [inline my-WOPI.py:8]: The suggested order for imports is to group your imports into the following categories:
* standard library imports
* other library imports
* imports from your local library
Also, each category should be alphabetical

So your imports would look like:
```
import argparse
import datetime
import json
from math import pi

import matplotlib
import matplotlib.animation as animation
import matplotlib.pyplot as plt
import numpy as np

```

File: my-WOPI.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+
+def parseCommands():
```
Comment (lbdreyer) [inline my-WOPI.py:11]: PEP8 prefers functions to be written in lowercase with underscores inbetween the words (see [PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names))
So here your function would be called `parse_commands`

File: my-WOPI.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+            patch_color = config['groups'][group]['color']
+            values = config['groups'][group]['data']
+            polygon_coords = np.array(list(zip(angles, values)))
+            polygon = matplotlib.patches.Polygon(polygon_coords, color=patch_color, alpha=0.4,
+                                                 label=config['groups'][group]['name'])
+            ax.add_patch(polygon)
+        ax.legend(loc='lower right')
+        plt.draw()
+
+    # initialise the first frame of the animation if animated, plot if otherwise
```
Comment (lbdreyer) [inline my-WOPI.py:251]: Comments like these can be really helpful! They help the person reading the code to understand what is going on
So don't be afraid to use them!

File: my-WOPI.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        plt.title(config['title'])
+        ax.set_rmax(config['max'])
+        ax.set_rmin(config['min'])
+
+        group_list = frame_list[i]
+        for group in group_list:
+            patch_color = config['groups'][group]['color']
+            values = config['groups'][group]['data']
+            polygon_coords = np.array(list(zip(angles, values)))
+            polygon = matplotlib.patches.Polygon(polygon_coords, color=patch_color, alpha=0.4,
```
Comment (lbdreyer) [inline my-WOPI.py:245]: I can't find an import for `matplotlib.patches`? It should ideally be added to your imports at the top of the file.

It may be that by importing matplotlib, it also imported matplotlib.patches under the hood, but you can't always rely on this! 
As an example, there was a submodule in iris that also imported when you would import iris. This happened because, at the time, it was being used in the `__init__.py` file of iris. However, a later release of iris removed the import of the submodule as it was no longer needed and that broke a bunch of people's code that were relying on this.

Basically, it is usually best practice to import every submodule you make use of.

File: my-WOPI.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        config (dict)
+    Returns:
+        list
+    """
+    frame_dict = dict()
+    for group, content in config['groups'].items():
+        frame_dict.setdefault(content['frame'], []).append(group)
+    ordered_frame_list = sorted(frame_dict)
+    min_ = ordered_frame_list[0]
+    max_ = ordered_frame_list[-1]
```
Comment (lbdreyer) [inline my-WOPI.py:214]: These are an example of when you could use names that explain more about what they contain.  So you could name these `min_frame` or perhaps `min_frame_number`.
You don't want names that are too long variable names. You just have to find a balance between length/meaningfulness.

File: my-WOPI.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    total_frame_list = [frame_dict[frame+min_] if frame+min_ in frame_dict else []
+                        for frame in range(max_-min_+1)]
```
Comment (lbdreyer) [inline my-WOPI.py:216]: I may be being a little nit-picky about this but I notice that this doesn't handle negative frame numbers.
So if you had frames like [-3, 0, 3, 5], this would produce a total_frame_list of [0, 1, 2, 3, 4, 5] 
Perhaps it would be worth catching that in the `cleanConfig` function?
I guess it depends on who would be writing the config files and whether they are likely to make mistakes like this that you need to catch.

File: my-WOPI.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+def main():
+    """Reads a config file specified in the command line and creates a radar plot from the data."""
+    args = parseCommands()
+    config = setConfig(args.config)
+    frame_list = orderFrames(config)
+
+    number_of_axes = len(config['axes'])
+    angle = 2*pi / number_of_axes
+    angles = [angle*x for x in range(number_of_axes)]
+    labels = config['axes']
```
Comment (lbdreyer) [inline my-WOPI.py:229]: You could have done before line 226 then reuse it when counting the number of axes, e.g.
axes_labels = config['axes']
number_of_axes = len(axes_lables)

File: my-WOPI.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    Args:
+
+    * config_file (string)
+
+    Returns:
+        dict
+    """
+    with open(config_file) as json_file:
+        config = json.load(json_file)
+    cleanConfig(config)
```
Comment (lbdreyer) [inline my-WOPI.py:40]: It's a common practice to order your functions such that they are defined above the place they are used. That means that iwere you to read through the module from start to end, when you get to a function, you would have already seen everything it uses.
So here you use `cleanConfig` but it is defined later on in the file, so it would be better to move this below `cleanConfig`

File: my-WOPI.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    return config
+
+
+def checkType(dict_, key, type_):
```
Comment (lbdreyer) [inline my-WOPI.py:44]: This name could be misleading, on first glance it may seem like a general type checking function, when it specifically look at values of the given dictionary. Perhaps could be renamed as `check_config_item_type` or if you want it to be more general to work with any dictionary you could do: `check_dict_item_type`

File: my-WOPI.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+    """
+    Raises appropriate exception if the key points to the wrong type of value
+
+    Args:
+
+    * dict_ (dict)
+    * key (string)
+    * type_ (type)
+    """
+    if type(dict_[key]) is not type_:
```
Comment (lbdreyer) [inline my-WOPI.py:54]: It is usually better practice to use `isinstance` as it is aware of inheritance of classes.

File: my-WOPI.py [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+        raise Exception("{0} does not have type {1}. {0} had type {2}".format(key, type_, type(dict_[key])))
+
+
+def cleanConfig(config):
```
Comment (lbdreyer) [inline my-WOPI.py:58]: This is quite a large function, you could look at breaking it down into smaller parts.
A problem with large functions, amongst others, is that they make unit testing more difficult.

File: my-WOPI.py [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+                frames_missing = True
+        else:
+            content['frame'] = 0
+    # pad the range of the plot appropriately
+    if 'max' not in config:
+        config['max'] = int(max(data_set) + 1)
+    if 'min' not in config:
+        config['min'] = int(min(data_set) - 1)
+        if config['min'] > 0:
+            config['min'] = 0
```
Comment (lbdreyer) [inline my-WOPI.py:129]: It doesn't seem necessary to include `min` in the config if you set it to be 0 anyway?
Unless you are also trying to handle negative data values, although I would have assumed that that would cause issues for your plots?

**Review by lbdreyer — COMMENTED** [review]
Overall a really great solution!
You managed to keep the implementation mostly quite simple. 

I have a few general comments:
* Good code structure
* Good use of docstrings
* *PEP8:* It's worth reading up about pep8, which is the style guide for Python. It helps make code readable by setting rules, such as a max line length of 79 characters, and naming conventions. there is also a command line tool called pycodestyle that you can run on your code to check that it is pep8 compliant.
* *Meaningful names:* Meaningful names for functions/variables can go a long way to helping other people to read and understand your code (and yourself if you have to go back and understand some code you wrote years ago and have forgotten how it all works). 

A good thing to keep in mind is that you spend more time reading code than writing it. Making your code easily readable and maintainable with the use of is really important.
